circleguard/investigator.py

```diff
@@ -226,42 +240,134 @@ def median_frametime(frametimes):
         return np.median(frametimes)
 
     @staticmethod
-    def _parse_beatmap(beatmap):
+    def _parse_beatmap(beatmap, easy, hard_rock):
         hitobjs = []
 
         # parse hitobj
-        for hit in beatmap.hit_objects_no_spinners:
-            p = hit.position
-            hitobjs.append([hit.time.total_seconds() * 1000, p.x, p.y])
+        for hit in beatmap.hit_objects(easy=easy, hard_rock=hard_rock):
+            hitobjs.append(hit)
         return hitobjs
 
     @staticmethod
-    def _parse_keypress_times(replay):
+    def _parse_replay(replay):
+        filtered_replay_data = []
         keypresses = replay.k & Investigator.MASK
         changes = keypresses & ~np.insert(keypresses[:-1], 0, 0)
-        return replay.t[changes != 0]
-
+        changes_t = np.diff(np.insert(replay.t, 0, 0))
+
+        for i, change in enumerate(changes):
+            if change != 0:
+                # not sure why, but notelock can be reduced by 1ms if the keypress frame has a frametime of 0ms
+                # so we need to keep track of the frames where this happens
+                if changes_t[i] == 0:
+                    filtered_replay_data.append([replay.t[i], replay.xy[i], True])
+                else:
+                    filtered_replay_data.append([replay.t[i], replay.xy[i], False])
+
+        # add a duplicate frame when 2 keys are pressed at the same time
+        changes = changes[changes != 0]
+        i = 0
+        for j in np.where(changes == 3)[0]:
+            filtered_replay_data.insert(j + i + 1, filtered_replay_data[j + i])
+            i += 1
+        
+        return np.array(filtered_replay_data, dtype=object)
```

Comment:
is it necessary to convert this to a numpy array?

Reply:
probably not